backstage/backstage
# Remove stale TechDocs artifacts after publishing new version

## 🧹  What / Why

This updates three publishers (GCS, AWS, Azure) in TechDocs common so that they delete stale files from storage during publish.  Prior to this, generated files based on deleted markdown would stick around in storage indefinitely. Same with images/assets/etc.

Closes #6132 

## How to Test

1. Generate a TechDocs site at least once that publishes to one of the updated storage providers.
2. Upload a file manually underneath the folder containing the generated docs that wasn't there previously
3. Re-generate the TechDocs site in order to re-trigger the publish.

You should see logs in your backend that indicate 1 stale file was deleted (and the file should be gone from storage).

![image](https://user-images.githubusercontent.com/6290749/128039322-dc7b43c9-1188-432e-98ce-f20d367e42d9.png)

## FYI
- We refactored some of the bulk storage operations into a helper in order to share concurrency/limiting logic.
- Listing files within a bucket under a folder is not a first-class operation in all storage providers, so we abstracted away some of this logic into helper methods on the publisher classes.
- We should consider refactoring the test/mock situation for these publishers sometime later.......

---
